### PR TITLE
cf-remote: Reuse SSH connections and other fixes (Digital Ocean Compatibility)

### PR DESCRIPTION
**Pull request by olehermanse**
The main reason why `install` wasn't working in Digital Ocean was that it opened a lot of SSH connections to the same host. This pull request fixes that and some other problems highlighted by LGTM code analysis.

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @vpodzime can review this?

**Comment by olehermanse** [comment]
@vpodzime With these changes, it seems to be working just fine for DO droplets (CentOS).

**Comment by olehermanse** [comment]
This pull request **introduces 5 alerts** and **fixes 2** when merging f9d5c30c2e3ec140185e2a25c2a2c0cd8a511d2b into 6f1e122b89c6ee91e1ef315f27932c73876122e7 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-bfbb602d396f73dc9d777599f7f596e7487ee4da)

**new alerts:**

* 5 for Explicit returns mixed with implicit \(fall through\) returns

**fixed alerts:**

* 2 for Cyclic import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by olehermanse** [comment]
This pull request **introduces 3 alerts** and **fixes 2** when merging a75c3d796e3f11673aecab377f316128a829a6fc into 6f1e122b89c6ee91e1ef315f27932c73876122e7 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-55b9bf28b1182b6a3720dd69ba96966ac03a7519)

**new alerts:**

* 2 for Explicit returns mixed with implicit \(fall through\) returns
* 1 for Unused import

**fixed alerts:**

* 2 for Cyclic import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by olehermanse** [comment]
This pull request **fixes 2 alerts** when merging 64cf033b897919a798d5cadc13f4e20eae08a4fd into 6f1e122b89c6ee91e1ef315f27932c73876122e7 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-1dd4e7dcadb67d86ca2e4dc45099e8681d444e76)

**fixed alerts:**

* 2 for Cyclic import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: contrib/cf-remote/cf_remote/ssh.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python3
```
Comment (vpodzime) [inline contrib/cf-remote/cf_remote/ssh.py]: I don't think a hash-bang is needed in this file.

File: contrib/cf-remote/cf_remote/ssh.py [hunk 2]
```diff
@@ -20,11 +22,25 @@ def connect(host, users=None):
     for user in users:
         try:
             c = fabric.Connection(host=host, user=user)
+
+            try:
+                c.open()
+            except (SSHException, ConnectionResetError, EOFError):
```
Comment (vpodzime) [inline contrib/cf-remote/cf_remote/ssh.py]: So the two new exceptions here are in the default Python's global scope?

Comment (olehermanse) [reply contrib/cf-remote/cf_remote/ssh.py]: Yes, both are from https://docs.python.org/3/library/exceptions.html

Comment (olehermanse) [reply contrib/cf-remote/cf_remote/ssh.py]: We could consider dropping this commit, I don't think it's needed after I made the other changes with reusing the connection.

File: contrib/cf-remote/cf_remote/ssh.py (hunk 1 above)
Comment (olehermanse) [reply contrib/cf-remote/cf_remote/ssh.py]: Thanks

File: contrib/cf-remote/cf_remote/ssh.py (hunk 2 above)
Comment (vpodzime) [reply contrib/cf-remote/cf_remote/ssh.py]: > We could consider dropping this commit, I don't think it's needed after I made the other changes with reusing the connection.

Cool, the fewer "random" exceptions we catch, the better.

**Comment by olehermanse** [comment]
This pull request **introduces 1 alert** and **fixes 2** when merging 04fde628a97ab28d5973656775a7ef1c7a1722d6 into 55d35ad6cbe306ba83c6d846f64ff247281f4465 - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-17bb5292bce12e866b420a91700e2d9f20bde50b)

**new alerts:**

* 1 for Explicit returns mixed with implicit \(fall through\) returns

**fixed alerts:**

* 2 for Cyclic import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**